CMakeLists.txt

```diff
@@ -1,72 +1,101 @@
 # ensmallen CMake configuration.  This project has no configurable options---it
 # just installs the headers to the install location, and optionally builds the
 # test program.
-cmake_minimum_required(VERSION 2.8.10)
-project(ensmallen C CXX)
+cmake_minimum_required(VERSION 3.3.2)
```

Comment:
If I recall correctly, Red Hat Enterprise Linux 7 (RHEL 7) has CMake 2.8.12 by default. RHEL 7 has gcc 4.8 by default, meaning it handles C++11 (using the -std=c++11 switch), and so it can compile ensmallen and mlpack.

RHEL 7 is used on a lot of computing clusters, suggesting that this CMake modernisation effort might end up reducing the uptake of ensmallen.

I generally don't see the point in bumping CMake requirements unless they bring in an important fix and/or functionality to the table.

On the Armadillo side I've made optional use of CMake 3.x features: https://gitlab.com/conradsnicta/armadillo-code/-/blob/9.850.x/CMakeLists.txt
Look for code with stuff like `if(NOT (${CMAKE_MAJOR_VERSION} LESS 3))`

Reply:
This caused me some amount of consternation also until I found the `cmake3` package in EPEL: https://rpmfind.net/linux/rpm2html/search.php?query=cmake3, so I think that should be sufficient for users on CentOS/RHEL 7.  I definitely agree that RHEL7 is a big target that we can't leave behind. :+1:

Reply:
I suggest expanding the README to inform folks on RHEL systems where the updated cmake package can be obtained.  For example:

> The installation of ensmallen requires cmake 3.3 or later.
> For older systems, such as RHEL 7, an updated version of cmake can be obtained from https://cmake.org or via the EPEL repository.